L1.md

```diff
@@ -0,0 +1,154 @@
+# C++ streaming coalescing API's
+
+* Author(s): ctiller
+* Approver: a11r
+* Status: Draft
+* Implemented in: C++
+* Last updated: 2017/01/13
+* Discussion at: https://groups.google.com/forum/#!topic/grpc-io/GfxB7nuV9GY
+
+## Abstract
+
+Provide C++ API's to coalesce:
+- initial metadata and first streaming message
+- final streaming message and trailing metadata
+into the same core batch.
+
+## Background
+
+Currently when making a streaming call using gRPC and sending a single message,
+three core batches are created: one for initial metadata, one for the message
+itself, and one for trailing metadata.
+
+Since the underlying transport has no knowledge that future messages are coming,
+it's forced to assume it should initiate writes immediately for each of them.
+In typical cases, this results in three sendmsg syscalls per stream.
+
+The current API allows some coalescing via WriteOptions. One can do:
+```c++
+auto s = stub->Foo(...);
+s->Write(first_message, WriteOptions().set_buffer_hint());
+s->WritesDone();
+```
+This is difficult to discover, and further more it performs two round-trips
+through the C Core stack, where it would be possible to do one.
+
+### Related Proposals:
+N/A
+
+## Proposal
+
+### Expand WriteOptions to include an end-of-stream bit
+```c++
+class WriteOptions {
+ public:
+  // ... existing interface ...
+
+  // last-message bit: indicates this is the last message in a stream
+  // client-side:  makes Write the equivalent of performing Write, WritesDone in
+  //               a single step
+  // server-side:  hold the Write until the service handler returns (sync api)
+  //               or until Finish is called (async api)
+  WriteOptions& set_last_message();
+  WriteOptions& clear_last_message();
+  bool is_last_message() const;
+};
+```
+
+### Add a convenience WriteLast method to ClientWriterInterface, ClientReaderWriterInterface, ServerWriterInterface, ServerReaderWriterInterface
```

Comment:
I am not sure we want all these additional convenience methods. The pro of course is that it these concepts are elevated to the top level in the API. But the con is that the API loses more orthogonality and there are to many ways of accomplishing the same thing.